### PR TITLE
Use a cache-backed client to reduce number of api requests

### PR DESCRIPTION
**Pull request by alvaroaleman**
We currently use an api reader which results in an actual api request
whenwever we need to get an object. This change makes us use the
default cache-backed client again. There are now a couple of conflict
errors in the log, but as that results in a retry it is not a
correctness issue.

Ref https://issues.redhat.com/browse/HOSTEDCP-154

### CODE REVIEW

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 1]
```diff
@@ -910,7 +910,7 @@ func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		}
 
 		awsCluster := controlplaneoperator.AWSCluster(controlPlaneNamespace.Name, hcluster.Name)
-		_, err = controllerutil.CreateOrPatch(ctx, r.Client, awsCluster, func() error {
+		_, err = r.CreateOrUpdate(ctx, r.Client, awsCluster, func() error {
```
Comment (alvaroaleman) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:913]: I changed this, because PATCH means that requests will never conflict which is not what we want, especially not if we increase the odds of having an outdated version because we use a cache.

Afterwards the tests started failing and it turns out that is because CreateOrPatch supports the status subresoruce whereas CreateOrUpdate does not, which is presumably the reason this was used in the first place. I've added support for the status subresource (through a hardcoded list) in the CreateOrUpdate provider and now everything works as exepcted :)

Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:913]: ok thanks, I created https://github.com/kubernetes-sigs/controller-runtime/issues/1725

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 2]
```diff
@@ -910,7 +910,7 @@ func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		}
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:911]: nit: should probably move this into its own commit.

Comment (alvaroaleman) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:911]: Moved it into a dedicated commit, ptal

**Comment by enxebre** [comment]
lgtm pending rebase

**Comment by alvaroaleman** [comment]
@enxebre @csrwng this now passed tests, ptal if you have time

File: support/upsert/upsert.go [hunk 3]
```diff
@@ -21,6 +24,12 @@ type CreateOrUpdateProvider interface {
 	CreateOrUpdate(ctx context.Context, c crclient.Client, obj crclient.Object, f controllerutil.MutateFn) (controllerutil.OperationResult, error)
 }
 
+var withStatusSubresource = sets.NewString(fmt.Sprintf("%T", &capiawsv1beta1.AWSCluster{}))
```
Comment (csrwng) [inline support/upsert/upsert.go:27]: Would a gvk make sense here? or is the list of types we'll support so limited that we don't care about name collisions?

Comment (alvaroaleman) [reply support/upsert/upsert.go:27]: I think packagename.Type is likely to remain unique, especially since the CAPI resources have the providername in the typename (i.E. its AWSCluster, not Cluster). Should I end up being wrong about this its easy to change later on
